File: 面试题29-顺时针打印矩阵.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        if(matrix.empty())
            return vector<int> {};

        int i1 = 0, j1 = 0, i2 = matrix.size() - 1, j2 = matrix[0].size() - 1;
        vector<int> result;

        while(i1 <= i2 && j1 <= j2){
            for(int k = j1; k < j2 + 1; k++){
                result.push_back(matrix[i1][k]);
            }

            for(int k = i1 + 1; k < i2; k++){  
                result.push_back(matrix[k][j2]);
            }

            if(i1 < i2){
                for(int k = j2; k >= i1; k--){
                    result.push_back(matrix[i2][k]);
                }
            }

            if(j1 < j2){
                for(int k = i2 - 1; k > i1; k--){
                    result.push_back(matrix[k][j1]);
                }
            }
            
            i1++;
            j1++;
            i2--;
            j2--;
        }

        return result;
    }
};

File: 面试题29-顺时针打印矩阵.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:

        def spiral_order(matrix, i1, j1, i2, j2):
            if i1 > i2 or j1 > j2:
                return []
            
            result = []
            result += matrix[i1][j1:j2 + 1]
            result += [matrix[k][j2] for k in range(i1 + 1, i2)]
            result += list(reversed(matrix[i2][j1:j2 + 1])) if i1 < i2 else []
            result += list(reversed([matrix[k][j1] for k in range(i1 + 1, i2)])) if j1 < j2 else []

            result += spiral_order(matrix, i1 + 1, j1 + 1, i2 - 1, j2 - 1)

            return result

        if not matrix:
            return []

        return spiral_order(matrix, 0, 0, len(matrix) - 1, len(matrix[0]) - 1)
